Add ~pred_pos~ `calc_animal_pos` function

damianooldoni opened the pull request:
This PR should solve #103.

To do:

- [x] add main funcionality as `pred_pos()` (snake case) and allow for flexibility in column names by adding opt args
- [x] add help function `predict_r` (snake_case) in `zzz.R` instead of `pred_pos.R` directly as it is used in one other function which could be added in the future, see [L842-L878](https://github.com/MarcusRowcliffe/CTtracking/blob/9601a79e2e66a4d0e6a831aeeb632e132f1d2c45/CTtracking.r#L842-L878)
- [x] add example input data as `./data/`
- [x] test all by writing unit-tests
- [x] add examples in documentation of the function

I made a tibble out of the animal data sent by @MarcusRowcliffe and adapted the code so that it runalso with tibbles, not only with basic data.frames. Still, if data.frames are passed as first arguments, the function works fine as it should. @MarcusRowcliffe: please test this with your own data to be even more sure.

damianooldoni commented:
@MarcusRowcliffe: could you give me a small `dat` file and a `mods` named list so that I can go further by writing all the tests and checking everything works fine? Of course, you final review will be requested once I have done with this.

peterdesmet commented on R/get_n_individuals.R:
```diff
@@ -21,8 +21,8 @@
 #' @param datapkg Deprecated. Use `package` instead.
 #' @param ... filter predicates for filtering on deployments
 #' @importFrom dplyr .data %>%
+#'
```
```suggestion
This
```

peterdesmet replied:
This was a test

MarcusRowcliffe commented:
>

damianooldoni commented:
I would like to solve #126 in this PR as well. Still, @peterdesmet  and @MarcusRowcliffe: you are able to review already.

peterdesmet commented on R/predict_pos.R:
```diff
@@ -0,0 +1,115 @@
+#'   positions for each digitised point. Default: `"x"`.
+#' @param y Character naming the column within `animal_pos` containing y pixel
+#'   positions for each digitised point. Default: `"y"`.
+#' @param image_width pixel x dimension of each image.
```
This and next: capitalize `Pixel`

damianooldoni replied:
Done while renaming in https://github.com/inbo/camtraptor/pull/124/commits/c127c42a5554d5a4611cd0121249fd34f2bb4f73

peterdesmet commented on R/data.R:
```diff
@@ -9,3 +10,14 @@
+#' It must contain at least columns containing the coordinates (x, y), the
+#' deployment and the image dimensions (width and height).
+"animal_positions"
+
```
In addition, the `animal_positions` data contains a number of columns we probably don't need, e.g.

- dir: C:/Users/rowcliffe.m/OneDrive - Zoological Society of London/GitHub/CTtracking/Survey_yyy/Deployments/S01
- DateTimeOriginal: 2017:10:02 19:06:43

@MarcusRowcliffe @damianooldoni can it be reduced to the columns that are relevant. Will help users understand what they need to provide as source data for this function.

damianooldoni replied:
GOod idea, @peterdesmet. If other functions need other columns we can import other very small dfs to handle it. I agree that these data examples should be as small as possible.

damianooldoni replied:
I did it for `animal_positions` but not for `dep_calib_models` as this list is actually an object of a specific class (`calibs`) (maybe created by @MarcusRowcliffe himself) and I don't know which fields are mandatory and which are not as I assume such calibs objects could be used for something else than just this.

damianooldoni replied:
To be followed in https://github.com/inbo/camtraptor/issues/135

peterdesmet commented on R/predict_pos.R:
```diff
@@ -0,0 +1,115 @@
+#'   additional columns:
+#'   - `radius`: radial distance from camera
+#'   - `angle`: angular distance from camera
+#'   - `frame_count`: indicator of the frame order within each sequence
```
This is a confusing name if it is not a count (sum). It is currently unclear to me what it is exactly or what it is based on.

damianooldoni replied:
It's the frame number within each sequence:

```r
output <- calc_animal_pos(animal_positions, dep_calib_models)
dplyr::select(output, c(sequenceID, frame_count))
print(dplyr::select(output, c(sequenceID, frame_count)), n = 20)
# A tibble: 42 × 2
   sequenceID frame_count
        <dbl>       <int>
 1          0           1
 2          0           2
 3          0           3
 4          0           4
 5          1           1
 6          1           2
 7          1           3
 8          1           4
 9          1           5
10          1           6
11          1           7
12          1           8
13          1           9
14          1          10
15          1          11
16          1          12
17         20           1
18         45           1
19         45           2
20         46           1
# … with 22 more rows
# ℹ Use `print(n = ...)` to see more rows
```

I propose: `frame_number`, but I am sure @peterdesmet you have better names in mind 😄

peterdesmet replied:
So it is the ordered number of the photo within a sequence? I'm assuming the function is not calculating that number, but it is provided in the original df? If so, I would not add it.

damianooldoni replied:
Actually there was a column called `frame_number` in the original data of @MarcusRowcliffe although not mandatory and so deleted while trimming unnecessary columns.

Here below how `frame_number `and `frame_count` look alike after running the function:

```r
output %>%
+   select(sequenceID, frame_number, frame_count) %>%
+   head(n = 20)
# A tibble: 20 × 3
   sequenceID frame_number frame_count
        <dbl>        <int>       <int>
 1          0            0           1
 2          0            1           2
 3          0            2           3
 4          0            3           4
 5          1            6           1
 6          1            7           2
 7          1            8           3
 8          1            9           4
 9          1           10           5
10          1           11           6
11          1           12           7
12          1           13           8
13          1           14           9
14          1           15          10
15          1           16          11
16          1           17          12
17         20          100           1
18         45           80           1
19         45           81           2
20         46           83           1
```

So, it's a kind of reset, starting to count from 1 for all sequences.

peterdesmet replied:
Ok, thanks. Imo, creating this column feels out of scope for this function, unless it is necessary for further user. It can easily be done by the user by sorting images on timestamp within a sequence. We had a similar discussion regarding the need for this field in Camtrap DP (https://github.com/tdwg/camtrap-dp/issues/10) and decided it can be easily created. Might be good to move this to an issue so that @MarcusRowcliffe can answer.

damianooldoni replied:
Thanks for posting the issue: https://github.com/inbo/camtraptor/issues/136.

peterdesmet reviewed (changes requested):
Seeing what the function does now, I think we should be clearer about the name. I think `predict` might be the wrong verb here? Why not `calculate`? And the position is calculated for (single) animals (only), so I think `animal` should be part of the name too. I would suggest:

`calc_animal_position()` or `calc_animal_pos()`

damianooldoni commented on R/data.R:
```diff
@@ -9,3 +10,14 @@
+#'
+#' It must contain at least columns containing the coordinates (x, y), the
+#' deployment and the image dimensions (width and height).
+"animal_positions"
```
Thanks to noticing it. I will remove them. I just dumped the original files of @MarcusRowcliffe here just to start working on and then I forgot removing them. Solved in https://github.com/inbo/camtraptor/pull/124/commits/bc54bb2e16ce0c1f315092d0658544683f4a8074

damianooldoni commented on R/predict_pos.R:
```diff
@@ -0,0 +1,115 @@
+#' @param calib_models Named list of deployment calibration models or site calibration
+#'   models (`calibs` objects), produced using `cal.site()` (not yet included in
+#'   this package). The deployment names are used as names.
+#' @param dep_tag Character naming the column within `animal_pos` against which
```
As I told you yesterday I tried to not change as less as possible. So, this part I didn't really changed. But you are right. We have to maintain same style as the rest of the package of course.

damianooldoni replied:
Solved in https://github.com/inbo/camtraptor/pull/124/commits/46e8ce43eb616613bc571838dae273982f59a034

damianooldoni commented on R/data.R:
```diff
@@ -9,3 +10,14 @@
+#' Sample of deployment calibration models or site calibration models (`depcal` objects)
+#' 
+#' A `calibs` object (list). The deployment names are used as names.
+"dep_calib_models"
```
Solved in https://github.com/inbo/camtraptor/pull/124/commits/d06e7039ab7e599b6b26aff5c9996abbdbf2b6ef

damianooldoni commented on R/get_n_individuals.R:
```diff
@@ -21,8 +21,8 @@
 #' @param datapkg Deprecated. Use `package` instead.
 #' @param ... filter predicates for filtering on deployments
 #' @importFrom dplyr .data %>%
```
Ehum, no. This file is full of `%>%` and `.data` as it relies on dplyr functions. Do you refer to another file? Or did you mean something different?

peterdesmet replied:
Well, for all other functions we use `pkg::function` but I understand that that is not workable for `.data` and `%>%`. I was just wondering if we can define it once for the whole package?

damianooldoni replied:
Well, once you have imported from at least one function, then you have it for all the other ones. Still, I don't find bad to have this in each function as it avoids breaking things if we remove the one containing the `@importFrom`

damianooldoni commented:
@peterdesmet: some updates:
1. I used snakeCase consistently in all default values and all column names in `animal_positions` loaded data.frame
2.  I added a new argument, `sequence_id` as it is required in the function body
3. I added a test checking the function works fine even if non default args (column names) are used
